holoviews/core/data/grid.py

```diff
@@ -491,7 +502,10 @@ def sample(cls, dataset, samples=[]):
                 data[d].append(arr)
             for vdim, array in zip(dataset.vdims, arrays):
                 flat_index = np.ravel_multi_index(tuple(int_inds)[::-1], array.shape)
-                data[vdim.name].append(array.flat[flat_index])
+                if da and isinstance(array, da.Array):
+                    data[vdim.name].append(array.flatten()[tuple(flat_index)])
```

Comment:
Think the indexing is the issue here. If you switch to `vindex`, IIUC what `flat_index` means, this should be lazy.

Reply:
ref: http://dask.pydata.org/en/latest/array-api.html#dask.array.Array.vindex

Reply:
Thanks, the same issue also applied in ``select`` which was causing most of the problems. Followed up in https://github.com/ioam/holoviews/pull/2424.